source/Irrlicht/COpenGLDriver.h

```diff
@@ -846,6 +848,38 @@ namespace video
             typedef std::pair<COpenGLVAOSpec::HashAttribs,COpenGLVAO*> HashVAOPair;
             HashVAOPair                 CurrentVAO;
             core::vector<HashVAOPair>    VAOMap;
+            // TODO this has to be peridically searched for shaders with refcount=1 and cleared
+            core::CObjectCache<core::smart_refctd_ptr<COpenGLSpecializedShader>, GLuint> ShaderProgramCache;
+
+            GLuint getShaderProgramGLname(COpenGLSpecializedShader* _shader)
+            {
+                auto rng = ShaderProgramCache.findRange(core::smart_refctd_ptr<COpenGLSpecializedShader>(_shader));
+                if (rng.first != rng.second)
+                    return rng.first->second;
+
+                //TODO

```

Comment:
and what do you think about this

Reply:
Actually COpenGLSpecializedShader objects are special, they can be shared.

So you can create and delete copies on different contexts than you're using, plus you can transfer the object from a dead context to a live context.

So technically speaking you don't really need a cache, just a mapping from live context to a shader slot (or some uint in range from 0 to `SCreationParameters::AxuGLContexts-1`) that will never change for a live context due to intialization and deinitialization of other contexts.

I used to have (in previous revisions) an ID assignment routine here to get this sort of mapping
https://github.com/buildaworldnet/IrrlichtBAW/blob/master/source/Irrlicht/COpenGLDriver.cpp#L465

Reply:
there might not be a need for such a cache after all, just an array of handles inside COpenGLSpecializedShader, that "hopefully" get a deferred intialization (if not you can create all copies at once)

Reply:
ok, it makes stuff easier. However theres going to be similar problems with VAOs, pipelines and FBOs. Also: for the ones aforementioned we have to make sure that once created GL object by thread A, gets deleted by thread A. So i think we might need per-thread/ctx refcounting inside some `COpenGL*` objects (pipeline, framebuffer). That way we could guarantee that GL object is being deleted in the same thread which created it as long as user makes proper amount of drops in relation to grabs. What troubles me is what if the last ref is held by cpu->gpu cache and most likely will be dropped in IAssetManager's destructor when app terminates..

Reply:
wait, when actually SAuxContext is being destroyed?

Reply:
Or rather: when SAuxContext::ctx is being destroyed?

Reply:
> wait, when actually SAuxContext is being destroyed?

When you call `COpenGLDriver::deinitAuxGLContexts` from the thread associated with `SAuxContexts`

Reply:
> So i think we might need per-thread/ctx refcounting inside some COpenGL* objects (pipeline, framebuffer). That way we could guarantee that GL object is being deleted in the same thread which created it as long as user makes proper amount of drops in relation to grabs. What troubles me is what if the last ref is held by cpu->gpu cache and most likely will be dropped in IAssetManager's destructor when app terminates..

For all cases but per-context shader copies.

This is why its best to cache `<GPUObjectHash,OpenGLHandle>` not `<GPUObjectPtr,OpenGLHandle>` because:
- you only care that the hash matches (multiple GPU objects can share the same OpenGL object)
- you don't care about mutability concerns (object changes = hash changes, you don't give a fuck)
- you don't care about lifetime of an object (object dies, caches object stays)

You can then periodically clear this cache according to LRU policy, you just set the max count of objects, if you get above that when retrieving an object, just delete the old ones + provide a function to clear these caches.

Reply:
yea, thats comfy. I wonder if hash for pipelines could be just 5 bits (present/absent shader stage) and treat it similarly to like VAOs are treated now (that is: 1) find VAO with matching active attribs and divisors; 2) rebind index and vertex buffers (modify GL vao state) ) ? That would mean max of 32 possible hash values (actually even less) so no need for any LRU cache

Reply:
no, it cant be like that. because VAO does not give a shit about which buffer is bound (cause they'll be rebound all the time).

Pipeline very much gives a shit about which shaders are bound (lots and lots of validation)

Reply:
using std::array<GLuint,5>